code/trans.c

```diff
@@ -0,0 +1,377 @@
+                       Ref new_list[],
+                       Count count)
+{
+  Res res;
+  Index i;
+  Count added = 0;
+
+  AVERT(Transform, transform);
+  AVER(old_list != NULL);
+  AVER(new_list != NULL);
+  /* count: cannot check */
+  
+  res = TableGrow(transform->oldToNew, count);
+  if (res != ResOK)
+    return res;
+
+  for(i = 0; i < count; ++i) {
+    /* NOTE: If the mutator isn't adding references while the arena is parked,
+       we might need to access the client-provided lists, using ArenaRead. */
+    if(old_list[i] == NULL)
+      continue;  /* permitted, but no transform to do */
+    if(old_list[i] == new_list[i])
+      continue;  /* ignore identity-transforms */
+
+    /* Old refs must be in managed memory. */
+    {
+      Seg seg;
+      AVER(SegOfAddr(&seg, transform->arena, old_list[i]));
+    }
+
+    res = TableDefine(transform->oldToNew, (Word)old_list[i], new_list[i]);
+    AVER(res != ResFAIL); /* It's a static error to add the same old twice. */
+    if (res != ResOK)
+      return res;
+    
+    ++added;
+  }
+
+  AVERT(Transform, transform);
+  
+  return ResOK;
+}
+
+
+/* TransformApply -- transform references on the heap */
+
+static Res transformFix(Seg seg, ScanState ss, Ref *refIO)
+{
+  Ref ref;
+  Transform transform;
+  Res res;
+
+  AVERT_CRITICAL(Seg, seg);
+  AVERT_CRITICAL(ScanState, ss);
+  AVER_CRITICAL(refIO != NULL);
+
+  transform = ss->fixClosure;
+  AVERT_CRITICAL(Transform, transform);
+
+  if (!transform->aborted) {
+    void *refNew;
+
+    ref = *refIO;
+
+    if (TableLookup(&refNew, transform->oldToNew, (Word)ref)) {
+      if (ss->rank == RankAMBIG) {
+        /* We rely on the fact that ambiguous references are fixed
+           first, so that no exact references have been transformed
+           yet. */
+        transform->aborted = TRUE;
+      } else {
+        /* NOTE: We could fix refNew in the table before copying it,
+           since any summaries etc. collected in the scan state will still
+           apply when it's copied.  That could save a few snap-outs. */
+        *refIO = refNew;
+      }
+    }
+  }
+
+  /* Now progress to a normal GC fix. */
+  /* TODO: Make a clean interface to this kind of dynamic binding. */
+  ss->fix = ss->arena->emergency ? SegFixEmergency : SegFix;
+  TRACE_SCAN_BEGIN(ss) {
+    res = TRACE_FIX12(ss, refIO);
+  } TRACE_SCAN_END(ss);
+  ss->fix = transformFix;
+
+  return res;
+}
+
+
+static void transformCondemn(void *closure, Word old, void *value)
+{
+  Seg seg;
+  GenDesc gen;
+  Bool b;
+  Trace trace = closure;
+
+  AVERT(Trace, trace);
+  UNUSED(value);
+
+  /* Find segment containing old address. */
+  b = SegOfAddr(&seg, trace->arena, (Ref)old);
+  AVER(b); /* old refs must be in managed memory, else client param error */
+
+  /* Condemn generation containing seg if not already condemned. */
+  gen = PoolSegPoolGen(SegPool(seg), seg)->gen;
+  AVERT(GenDesc, gen);
+  if (RingIsSingle(&gen->trace[trace->ti].traceRing))
```

Comment:
Use `TraceSetIsMember(gen->activeTraces, trace)` here to avoid needing to know so many details of the data structures.

Reply:
JPH todo